Branch-SDK/src/io/branch/referral/ExtendedAnswerProvider.java

```diff
@@ -0,0 +1,152 @@
+
+import android.text.TextUtils;
+import android.util.Log;
+
+import com.crashlytics.sdk.android.answersshim.AnswersOptionalLogger;
+import com.crashlytics.sdk.android.answersshim.KitEvent;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.Iterator;
+
+/**
+ * Created by sojanpr on 6/8/16.
+ * <p>
+ * Class for providing extended data to Fabric Answers.
+ * Create Kit events for the "install" / "open" / "share" events from Branch and update the Answers through
+ * Answers-Shim API. Class handles the logic for flattening  a Branch request to discrete key-value pairs for Answer {@link KitEvent}.
+ * </p>
+ */
+class ExtendedAnswerProvider {
+    // Event Name for Kit Event
+    private static final String KIT_EVENT_INSTALL = "install";
+    private static final String KIT_EVENT_OPEN = "open";
+    private static final String KIT_EVENT_SHARE = "share";
+
+    private static final String EXTRA_PARAM_NOTATION = "+";
+    private static final String CTRL_PARAM_NOTATION = "~";
+    private static final String INNER_PARAM_NOTATION = ".";
+
+    /**
+     * <p>
+     * Method for sending the kit events to Answers for Branch Events.
+     * Method create a Json Object to flatten and create the {@link KitEvent} with key-values
+     * </p>
+     *
+     * @param request  {@link ServerRequest} corresponding to the event
+     * @param response {@link ServerResponse} corresponding to the event
+     */
+    public static void provideData(ServerRequest request, ServerResponse response) {
+        JSONObject jsonData = null;
+        ExtendedKitEvent kitEvent = null;
+        try {
+            if (request instanceof ServerRequestInitSession) {
+                if (response.getObject() != null && response.getObject().has(Defines.Jsonkey.Data.getKey())) {
+                    if (request instanceof ServerRequestRegisterInstall) {
+                        kitEvent = new ExtendedKitEvent(KIT_EVENT_INSTALL);
+                    } else {
+                        kitEvent = new ExtendedKitEvent(KIT_EVENT_OPEN);
+                    }
+                    jsonData = new JSONObject();
+                    jsonData.put(Defines.Jsonkey.Data.getKey(), new JSONObject(response.getObject().getString(Defines.Jsonkey.Data.getKey())));
+                }
+            } else if (request instanceof ServerRequestCreateUrl && ((ServerRequestCreateUrl) request).isReqStartedFromBranchShareSheet()) {
+                BranchLinkData linkData = ((ServerRequestCreateUrl) request).getLinkPost();
+                if (linkData != null && linkData.has(Defines.Jsonkey.Data.getKey())) {
+                    kitEvent = new ExtendedKitEvent(KIT_EVENT_SHARE);
+                    jsonData = new JSONObject();
+                    JSONObject linkDataJson = new JSONObject(linkData.getString(Defines.Jsonkey.Data.getKey()));
+
+                    jsonData.put(Defines.LinkParam.Channel.getKey(), linkData.getChannel());
```

Comment:
Is there anyway to iterate through the data rather than hard code all these specific calls? I'd err on sending more data than not. See Ahmed's implementation.